loopy/transform/reindexing.py

```diff
@@ -0,0 +1,339 @@
+        from pymbolic.mapper.substitutor import make_subst_func
+        from pymbolic.primitives import Subscript, Variable
+
+        rec_indices = tuple(self.rec(idx, expn_state) for idx in expr.index_tuple)
+
+        assert len(self.index_names) == len(rec_indices)
+        subst_func = make_subst_func({idx_name: rec_idx
+                                      for idx_name, rec_idx in zip(self.index_names,
+                                                                   rec_indices)})
+
+        return SubstitutionMapper(subst_func)(
+            Subscript(Variable(self.reindexed_var_name),
+                      self.new_index_expr)
+        )
+
+
+def reindex_using_seghir_loechner_scheme(kernel: LoopKernel,
+                                         var_name: str,
+                                         *,
+                                         i_agree_to_the_gpl_terms: bool = False
+                                         ) -> LoopKernel:
+    """
+    Returns a kernel with expressions of the form ``var_name[i1, .., in]``
+    replaced with ``var_name_reindexed[f(i1, .., in)]`` where ``f`` is a
+    quasi-polynomial as outlined in [Seghir_2006]_.
+
+    :arg i_agree_to_the_gpl_terms: If *False*, prints a warning informing the
+        caller that the returned kernel object would be a derivative of GPL
+        licensed work.
+    """
+    from loopy.transform.subst import expand_subst
+    from loopy.symbolic import (get_access_map, pw_qpolynomial_to_expr,
+                                SubstitutionRuleMappingContext)
+
+    if var_name not in kernel.temporary_variables:
+        raise LoopyError(f"'{var_name}' not in temporary variable in kernel"
+                         f" '{kernel.name}'.")
+
+    # {{{ compute the access_range of *var_name* in *kernel*
+
+    subst_kernel = expand_subst(kernel)
+    vng = kernel.get_var_name_generator()
+    new_var_name = vng(var_name+"_reindexed")
+
+    access_ranges: List[ISLSetT] = []
+
+    for insn in subst_kernel.instructions:
+        domain = subst_kernel.get_inames_domain(insn.within_inames)
+        if isinstance(insn, MultiAssignmentBase):
+            access_exprs = _IndexCollector(var_name)((insn.assignees,
+                                                      insn.expression,
+                                                      tuple(insn.predicates)))
+        elif isinstance(insn, (CInstruction, BarrierInstruction)):
+            access_exprs = _IndexCollector(var_name)(insn.predicates)
+        else:
+            raise NotImplementedError(type(insn))
+
+        for access_expr in access_exprs:
+            access_ranges.append(get_access_map(domain,
+                                                access_expr.index_tuple,
+                                                assumptions=kernel.assumptions)
+                                 .range())
+
+    del subst_kernel
+
+    # }}}
+
+    access_range = _union_access_ranges(access_ranges)
```

Comment:
Is the existing access range mapper not suitable? Could it be made suitable?

Reply:
Aah very convenient. Yep, hadn't carefully looked into using it. Simply used it. (Cut down the source code lines by ~30 lines) Thanks!